File: dynamic_programming/best_time_to_buy_and_sell_stock_with_cooldown.py
def maxProfitDS1(prices: list[int]) -> int:
    sold, hold, reset = float('-inf'), float('-inf'), 0

    for price in prices:
        # Alternative: the calculation is done in parallel
        # Therefore, there is no need to keep temporary variables
        # sold,hold,reset = hold + price,max(hold,reset - price),max(reset,sold)

        pre_sold = sold
        sold = hold + price
        hold = max(hold, reset - price)
        reset = max(reset, pre_sold)

    return max(sold, reset)


def maxProfitDS2(prices: list[int]) -> int:
    num_days = len(prices)

    # Pad array with additional zero to simplify the logic
    max_profit = [0] * (num_days + 2)

    for prev_day in range(num_days - 1, -1, -1):
        case1 = 0

        # Case 1) Buy and sell the stock
        for curr_day in range(prev_day + 1, num_days):
            profit = ((prices[curr_day] - prices[prev_day])
                      + max_profit[curr_day + 2])

            case1 = max(profit, case1)

        # Case 2) No transaction with the stock at prices[prev_day]
        case2 = max_profit[prev_day + 1]

        # Wrap up the two cases
        max_profit[prev_day] = max(case1, case2)

    return max_profit[0]

File: dynamic_programming/best_time_to_buy_and_sell_stock_with_cooldown.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfitDS1 = function(prices) {
    let sold = Number.MIN_SAFE_INTEGER;
    let hold = Number.MIN_SAFE_INTEGER;
    let reset = 0;

    for (const price of prices)
    {
        const pre_sold = sold;
        sold = hold + price;
        hold = Math.max(hold, reset - price);
        reset = Math.max(reset, pre_sold);
    }

    return Math.max(sold, reset);
};

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfitDS2 = function(prices) {
    let max_profit = new Array(prices.length + 2).fill(0);

    for (let prev_day = prices.length - 1; prev_day >= 0; prev_day--)
    {
        let case1 = 0;

        for (let curr_day = prev_day + 1; curr_day < prices.length; curr_day++)
        {
            const profit =
                prices[curr_day] - prices[prev_day] + max_profit[curr_day + 2];
            case1 = Math.max(profit, case1);
        }

        const case2 = max_profit[prev_day + 1];

        max_profit[prev_day] = Math.max(case1, case2);
    }

    return max_profit[0];
};
